### PR TITLE
feat: builder's guide and dapp overview stub

### PR DESCRIPTION
**Pull request by cha0sg0d**

### CODE REVIEW

File: builders_guide.md [hunk 1]
```diff
@@ -0,0 +1,547 @@
...
+- What is [The Graph](https://thegraph.com/) and why is it useful?
+    - [The Graph](https://thegraph.com/) is an indexing protocol for querying networks like Ethereum and IPFS. Anyone can build and publish open APIs, called subgraphs, making data easily accessible.
+    - The graph enables more complex queries like who owned this planet at time *t* and has a much faster performance than a vanilla JSON-RPC request.
+- The amazing Jacob Rosenthal has created a DF subgraph that can be deployed via just a few commands. Follow his instructions [here](https://github.com/darkforest-eth/eth/tree/master/subgraph) to deploy the subgraph.
+    - *Note: If you’re on a Mac M1, Docker might not work, and you won’t be able to test the subgraph locally.*
+- If you are changing a fundamental data structure of the subgraph, you will need to update the schema and mappings. See [here](https://github.com/cha0sg0d/eth/commit/1b382624c986ed2b045b8465bde2df543dc5d3f0) for an example edit.
+
+## Side Quest 2: Auto Whitelisting with no server.
+
+*Warning: This information is potentially out of date*
```
Comment (phated) [inline builders_guide.md:512]:
```suggestion
```

It was, but let's fix it.

File: builders_guide.md [hunk 2]
```diff
@@ -0,0 +1,547 @@
...
+
+- Get a csv file of all the addresses you want to add to the game.
+- The following command assumes you already have deployed your contracts to production
+- Run:
+    - `yarn workspace eth hardhat:prod whitelist-nokey:register --path <csv of addresses>`
```
Comment (phated) [inline builders_guide.md:517]:
```suggestion
    - `yarn workspace eth hardhat:prod whitelist:registerAddress --address < cat address.csv`
```
Something like this? it looks like that task takes a comma separated list of values

File: builders_guide.md [hunk 3]
```diff
@@ -0,0 +1,547 @@
...
+
+## Side Quest 3: Publishing your new contracts as an npm package.
```
Comment (phated) [inline builders_guide.md:519]:
```suggestion
## Side Quest 3: Publishing your new contracts as an npm package
```

File: builders_guide.md [hunk 4]
```diff
@@ -0,0 +1,547 @@
...
+
+- Will update more later.
```
Comment (phated) [inline builders_guide.md:521]:
```suggestion
- Work in progress...
```

File: builders_guide.md [hunk 5]
```diff
@@ -0,0 +1,547 @@
...
+
+
+## Troubleshooting
+
+- If you see any zk checks fail, like `failed init proof check`, you need to recompile the circuits with the following command:
```
Comment (phated) [inline builders_guide.md:526]:
```suggestion
- If you see any ZK checks fail, like `failed init proof check`, you need to recompile the circuits with the following command:
```

File: builders_guide.md [hunk 6]
```diff
@@ -0,0 +1,547 @@
...
+    - `yarn workspace eth circom:dev`
+    - Wait ~5 minutes for the circuits to recompile. This will result in a new `Verifier.sol` file.
+- If you see the following warning
+    
+    > Warning: Contract code size is 24622 bytes and exceeds 24576 bytes (a limit introduced in Spurious Dragon). This contract may not be deployable on mainnet. Consider enabling the optimizer (with a low "runs" value!), turning off revert strings, or using libraries.
+    > 
+    - You need to reduce the contract size. Try commenting out changeAdmin or buyHat functions in DarkForestCore. Note that if you do, you will also have to comment out or delete the tests that use these functions.
+
+## Appendix 1: Using Git effectively.
```
Comment (phated) [inline builders_guide.md:535]:
```suggestion
## Appendix 1: Using Git effectively
```

File: builders_guide.md [hunk 7]
```diff
@@ -0,0 +1,547 @@
...
+
+- For setting up your environment initially, see the [Better Method for Running a Local Game](README.md#better-method-for-running-a-local-game)
```
Comment (phated) [inline builders_guide.md:537]: Need to update this link, it has the old anchor (and title)

File: dapp.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
+# Dark Forest Dapp Overview
+
+This page will provide a high level overview of the components and structure of the Dark Forest codebase. 
+
+*This page is not official in any way and is subject to change as the game evolves.*
+
+# Intro
+- Dark Forest has a lot of moving parts, but at its core, it is a blockchain-based application, commonly referred as a [dapp](https://ethereum.org/en/developers/docs/dapps/).
+    
+    > *A dapp has its backend code running on a decentralized peer-to-peer network (blockchain). Contrast this with an app where the backend code is running on centralized servers.   A dapp can have frontend code and user interfaces written in any language (just like an app) to make calls to its backend*
```
Comment (phated) [inline dapp.md:10]:
```suggestion
    > A dapp has its backend code running on a decentralized peer-to-peer network (blockchain). Contrast this with an app where the backend code is running on centralized servers. A dapp can have frontend code and user interfaces written in any language (just like an app) to make calls to its backend
```

File: dapp.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    > 
+- Dark Forest has a backend that contains the **[code](https://github.com/darkforest-eth/eth/)** that enforces the game physics and rules. This code is divided into files called **[smart contracts](https://github.com/darkforest-eth/eth/tree/master/contracts),** and it runs on a blockchain.
```
Comment (phated) [inline dapp.md:12]:
```suggestion
- Dark Forest has a backend that contains the **[code](https://github.com/darkforest-eth/eth/)** that enforces the game constraints and rules. This code is divided into files called **[smart contracts](https://github.com/darkforest-eth/eth/tree/master/contracts),** and it runs on a blockchain.
```
Still unsure if constraints is the best term

**Review by phated — CHANGES_REQUESTED** [review]
Looking really good! Most of my comments are just nits, but I think we should adjust some of the code snippets for R4 (using `yup` instead of `decoders`)

File: builders_guide.md [hunk 10]
```diff
@@ -0,0 +1,547 @@
...
+
+## 2.0 Understand `darkforest.toml`
+
+- Now we are going to make some changes to the game via the settings file. This file is called `darkforest.toml` and lives in the [eth](https://github.com/darkforest-eth/eth/) repository.
+- The `darkforest.toml` is useful because it allows you to configure a large number game parameters without writing a single line of code. Understanding what each parameter means is also a great way to get acquainted with the codebase.
+- Read the explanation of each setting in this [gist](https://gist.github.com/cha0sg0d/2d4d9b95a5f90c1b0e285b601b850827).
+
+## 2.1 Configure `darkforest.toml`
+
+- Add the following to the `darkforest.toml` file.
```
[the remark this answers is not in the document]
Comment (cha0sg0d) [reply builders_guide.md:85]: Will this .toml get used for a localhost game automatically?

[the remark this answers is not in the document]
Comment (phated) [reply builders_guide.md:85]: Yep, the settings loader will prefer a toml file that includes your network name over the one without. See https://github.com/darkforest-eth/eth/blob/master/settings.ts#L260

File: builders_guide.md [hunk 11]
```diff
@@ -0,0 +1,547 @@
...
+    
+    ```typescript
+    // ~ line 162
+    UPGRADEABLE_PLANETS: withDefault<ExactArray5<boolean>>(
+    		exactArray5(decoders.boolean),
+    		[true,false,false,false,false]
+    )
+    ```
+    
+- Run `yarn hardhat compile` to make sure that the contract compiles successfully.
```
[the remark this answers is not in the document]
Comment (cha0sg0d) [reply builders_guide.md:175]: Typo on workspace

[the remark this answers is not in the document]
Comment (cha0sg0d) [reply builders_guide.md:175]: I'll fix it

File: builders_guide.md [hunk 12]
```diff
@@ -0,0 +1,547 @@
...
+    - This functionality is built-in to the local deploy process, but needs to happen manually for deploying to production.
+
+
+## Side Quest 1: Deploying the subgraph
+
+- What is [The Graph](https://thegraph.com/) and why is it useful?
+    - [The Graph](https://thegraph.com/) is an indexing protocol for querying networks like Ethereum and IPFS. Anyone can build and publish open APIs, called subgraphs, making data easily accessible.
+    - The graph enables more complex queries like who owned this planet at time *t* and has a much faster performance than a vanilla JSON-RPC request.
+- The amazing Jacob Rosenthal has created a DF subgraph that can be deployed via just a few commands. Follow his instructions [here](https://github.com/darkforest-eth/eth/tree/master/subgraph) to deploy the subgraph.
+    - *Note: If you’re on a Mac M1, Docker might not work, and you won’t be able to test the subgraph locally.*
```
[the remark this answers is not in the document]
Comment (cha0sg0d) [reply builders_guide.md:507]: I had trouble when I tested it in January, looks like it's fixed now.

File: builders_guide.md (hunk 1 above)
Comment (cha0sg0d) [reply builders_guide.md:512]: I will link to my Whitelist.sol and hardhat test if people want to set up no server whitelisting.

File: builders_guide.md (hunk 2 above)
Comment (cha0sg0d) [reply builders_guide.md:517]: ooh that's a nice way of doing it

File: builders_guide.md (hunk 7 above)
Comment (cha0sg0d) [reply builders_guide.md:537]: will do

File: builders_guide.md [hunk 13]
```diff
@@ -0,0 +1,547 @@
...
+    })
+    
+    ```
+    
+- In `settings.ts`:
+    
+    ```typescript
+    // ~ line 162
+    UPGRADEABLE_PLANETS: withDefault<ExactArray5<boolean>>(
+    		exactArray5(decoders.boolean),
```
[the remark this answers is not in the document]
Comment (cha0sg0d) [reply builders_guide.md:169]: doing it rn

File: builders_guide.md [hunk 14]
```diff
@@ -0,0 +1,547 @@
...
+    ```
+    
+- Add to `eth/test/utils/TestWorld.ts`:
+    
+    ```typescript
+    import { .., customUpgradeInitializers } from './WorldConstants';
+    
+    export function customUpgradeWorldFixture(): Promise<World> {
+      return initializeWorld({
+        //@ts-expect-error due to boolean[] vs ExactArray<boolean> types.
```
[the remark this answers is not in the document]
Comment (cha0sg0d) [reply builders_guide.md:199]: confirm goes away

File: builders_guide.md [hunk 15]
```diff
@@ -0,0 +1,547 @@
...
+
+- Same as 1.1, but now you’re in a custom universe that you designed!
+- Without writing even one line of code, you’ve just made an awesome mode of Dark Forest. Next, we are going to learn how to modify the game even further.
+
+# Quest 3: Local Game Custom Client and Contracts.
+
+- For this quest, you will create a Dark Forest game with the following custom rule change: All planets can be upgraded, not just planets of type PLANET.
+	- If you are new to Dark Forest, there are 5 planet types: Planet, Asteroid, Foundry, Spacetime Rip, and Quasar. Currently, only the Planet type can be upgraded to improve its defense, range, or speed. 
+- We will add a new setting, `UPGRADEABLE_PLANETS` that is an array of 5 booleans. If an element is true, the corresponding planet type can be upgraded. If an element is false, the corresponding planet type cannot be upgraded.
+- Before you get started, review the [structure](dapp.md) of the Dark Forest codebase.
```
[the remark this answers is not in the document]
Comment (cha0sg0d) [reply builders_guide.md:119]: added link to client overview

File: builders_guide.md [hunk 16]
```diff
@@ -0,0 +1,551 @@
...
+
+- Now we are going to make some changes to the game via the settings file. This file is called `darkforest.toml` and lives in the [eth](https://github.com/darkforest-eth/eth/) repository.
+- The `darkforest.toml` is useful because it allows you to configure a large number game parameters without writing a single line of code. Understanding what each parameter means is also a great way to get acquainted with the codebase.
+- Read the explanation of each setting in this [gist](https://gist.github.com/cha0sg0d/2d4d9b95a5f90c1b0e285b601b850827).
+  - *Note: Parameters that affect shrinking are not in the base codebase, so can't be used for this tutorial*.
+
+## 2.1 Configure `darkforest.toml`
+
+- Duplicate the existing the `darkforest.toml` file.
+- Rename it `darkforest.local.toml`
```
Comment (phated) [inline builders_guide.md:87]:
```suggestion
- Rename it `darkforest.localhost.toml`
```
It uses the full network name, which I believe is always `localhost` in the hardhat node scenario.

File: builders_guide.md [hunk 17]
```diff
@@ -0,0 +1,551 @@
...
+  - *Note: The settings loader will [prefer](https://github.com/darkforest-eth/eth/blob/master/settings.ts#L260) a toml file that includes your network name over the one without.*
+  - *If you want to deploy your .toml changes to production, put them in `darkforest.xdai.toml`.*
+
+- Add the following to `darkforest.local.toml`:
```
Comment (phated) [inline builders_guide.md:91]:
```suggestion
- Add the following to `darkforest.localhost.toml`:
```
It uses the full network name, which I believe is always `localhost` in the hardhat node scenario.

File: builders_guide.md [hunk 18]
```diff
@@ -0,0 +1,551 @@
...
+- It’s super fun.
+
+# Guide Structure
+
+## Main Quests
+
+0. [Download the correct code and install dependencies](#quest-0-installation-and-setup)
+1. [Deploy a local game](#quest-1-local-game)
+2. [Deploy a local game with a unique configuration via `darkforest.toml`](#quest-2-local-game-with-custom-configuration)
+3. [Deploy a local game with a modified client and contracts](#quest-3-local-game-custom-client-and-contracts)
```
Comment (phated) [inline builders_guide.md:21]:
```suggestion
3. [Deploy a local game with a modified client and contracts](#quest-3-local-game-with-custom-client-and-contracts)
```
Testing out the links and this one changed.

File: builders_guide.md [hunk 19]
```diff
@@ -0,0 +1,551 @@
...
+
+- Same as 1.2, but check to see that there is a level 9 planet at location (0,0) when you join the universe.
+- The existing deploy script should automatically create the level 9 planet because it calls the Hardhat `createPlanets` task.
+
+## 2.5 Play your game
+
+- Same as 1.1, but now you’re in a custom universe that you designed!
+- Without writing even one line of code, you’ve just made a custom mode of Dark Forest. Next, we are going to learn how to modify the game even further.
+
+# Quest 3: Local game with custom client and contracts.
```
Comment (phated) [inline builders_guide.md:118]:
```suggestion
# Quest 3: Local game with custom client and contracts
```

File: builders_guide.md [hunk 20]
```diff
@@ -0,0 +1,551 @@
...
+- Before you get started, review the [structure](dapp.md#client-overview) of the Dark Forest codebase.
+
+## 3.0 Set up a new config variable
+
+- We want future users to be able to control planet upgrades from the `darkforest.toml` file. This allows for easy customization of the game for non-technical users.
+
+### 3.0.1 Add a new `UPGRADEABLE_PLANETS` variable to the contracts
+
+- With these edits, we are simply adding a new constant to the codebase that will be set before the contracts are deployed in the `darkforest.toml`. To have a variable reflected across the smart contracts requires making changes in a few places:
+- In `darkforest.toml`:
```
Comment (phated) [inline builders_guide.md:132]:
```suggestion
- With these edits, we are simply adding a new constant to the codebase that will be set before the contracts are deployed in the `darkforest.localhost.toml`. To have a variable reflected across the smart contracts requires making changes in a few places:
- In `darkforest.localhost.toml`:
```

File: builders_guide.md [hunk 21]
```diff
@@ -0,0 +1,551 @@
...
+    }
+    ```
+    
+
+## 3.7 Show user which planets are upgradeable
+
+- In `client/src/Frontend/Panes/UpgradeDetailsPane.ts`:
+    - Try to change this logic yourself to allow only the planets marked as true in `UPGRADEABLE_PLANETS` variable to be upgraded in the client.
+    - Hint: This is very similar to the change we made to `DarkForestPlanet.sol` but in TypeScript.
+    - Also, to get access to the `UPGRADEABLE_PLANETS` value, we need to access the GameManager object. The React [Frontend](client/src/Frontend) doesn’t have direct access to the GameManager, but it does have access to the GameUiManager, which can fetch the GameManager.
```
Comment (phated) [inline builders_guide.md:452]:
```suggestion
    - Also, to get access to the `UPGRADEABLE_PLANETS` value, we need to access the GameManager object. The React [Frontend](https://github.com/darkforest-eth/client/tree/master/src/Frontend) doesn’t have direct access to the GameManager, but it does have access to the GameUiManager, which can fetch the GameManager.
```
It seems like we can't link into the submodules from github UI, so let's direct to the repo

File: builders_guide.md [hunk 22]
```diff
@@ -0,0 +1,551 @@
...
+</details>
+        
+
+## 3.8 Run a local game
+
+- `yarn` to compile the client side changes
+- `yarn start` to launch the game. (Same as step 1.1)
+    - *Note: For test purposes, you can set `DISABLE_ZK_CHECKS` to true for the local game.*
+- See if you can upgrade an asteroid field!
+    - [Video showing upgradable asteroid](https://twitter.com/cha0sg0d_/status/1484620387866447873)
```
Comment (phated) [inline builders_guide.md:485]:
```suggestion
    - [It should look something like this](https://twitter.com/cha0sg0d_/status/1484620387866447873)
```
Thought this was alt-text. Probably want to say it better.

File: builders_guide.md [hunk 23]
```diff
@@ -0,0 +1,551 @@
...
+- For setting up your environment initially, see the [If you plan to make changes to darkforest-local](README.md#if-you-plan-to-make-changes-to-darkforest-local) steps.
+- Each time you want to add a new feature, make a new branch on that specific repository.
+    - `git checkout -b <new_feature>`
+- When the feature is ready to go, merge it with your main branch and delete the feature branch.
+    - `git checkout main`
+    - `git merge new-feature`
+    - `git branch -d new-feature`
+- When making changes to the codebase, describe each change you make via `git add` and `git commit -m <your_messsage>`
+- It is crucial that each edit is understandable to someone who did not write this code.
+- See my `darkforest/eth` [repo](https://github.com/cha0sg0d/eth/commits/community) for an example of documenting each commit.
```
Comment (phated) [inline builders_guide.md:550]:
```suggestion
- See [cha0sg0d's](https://github.com/cha0sg0d) `darkforest/eth` [repo](https://github.com/cha0sg0d/eth/commits/community) for an example of documenting each commit.
```

**Review by phated — CHANGES_REQUESTED** [review]
Just a few things left

**Review by phated — APPROVED** [review]
Amazing work! Excited to land this 🎉

**Pull request merged**